doc/api/util.md

```diff
@@ -1020,6 +1020,81 @@ Otherwise, returns `false`.
 See [`assert.deepStrictEqual()`][] for more information about deep strict
 equality.
 
+## `util.parseArgs([config])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+* `config` {Object} Used to provide arguments for parsing and to configure
+  the parser. `config` supports the following properties:
+  * `args` {string\[]} array of argument strings. **Default:** `process.argv`
+    with `execPath` and `filename` removed.
```

Comment:
Is there a reason why this allows users to configure the list of arguments? I was wondering if parseArgs can take care of all use cases where manual handling of the list of arguments is involved.

Reply:
This is intended as an MVP; there’s far too many such use cases to be able to handle them all without 2x or 3x the complexity.

Reply:
Are we willing to make this complex enough in further iterations to handle all such use cases?

Reply:
I certainly am willing to make it more complex in further iterations to handle reasonable use cases; i think it’d be unwise to commit to “all”.

Reply:
Just as an example: almost all CLI applications treat `--` as signaling the end of flags, such that every following argument should be treated as positional.  So `ls -- -l` lists the directory `-l`. And `ls -- .` lists the current directory, just like `ls .` does.

But some small fraction of applications, like `npm`, have different behavior for `--`, typically slicing the arguments following `--` and passing them to another command. Right now, the way you'd do that with `parseArgs` is

```js
let dashDashIndex = process.mainArgs.indexOf('--');
let mainArgs = dashDashIndex === -1 ? process.mainArgs : process.mainArgs.slice(0, dashDashIndex);
let forwarded = dashDashIndex === -1 ? [] : process.mainArgs.slice(dashDashIndex);
let mainCommand = parseArgs({ config, args: mainArgs });
```

which is, I think, relatively straightforward and clearly expresses what you're trying to do. This is a pretty rare use case, and it's pretty easy to handle _if_ `args` is configurable. I think it's good that this use case is supported, but I also don't think it would make sense to have explicit configuration which changed how `--` works just so that this rare use case can avoid this small amount of work. It's better to keep `parseArgs` simple.

So I don't think we should have as a goal "handling all such cases". Complexity is bad, and it's OK if some use cases require the user to do a bit of extra imperative work rather than `parseArgs` having a tunable declarative configuration for every possible use case.

Reply:
> Is there a reason why this allows users to configure the list of arguments? 

Hopefully typical use can rely on the implicit argument detection, but there are multiple reasons the author may wish to supply the arguments explicitly:

1) unit testing. The CLI is the public interface of a CLI program and the author may want to validate the parsing works as intended. (i.e. without monkeying with `process.argv`.)

2) Situations where want parsing of arguments not supplied via node CLI arguments. Examples that have come up for Commander include a web connection, and a chat bot.

3) Custom preprocessing of arguments to extend or modify what `parseArgs` offers. Like custom `--` behaviour in previous comment, or alias expanding `-ws` to `--workspace default` say.

4) Supplying arguments in cases where the runtime environment means the user supplied arguments do not start at usual position. Example: [zx](https://github.com/google/zx)

Reply:
5. trying to retain backwards compatibility with a legacy API that wouldn't be reasonable to support in core.

Reply:
I won't reiterate other people's points, but I think there's a lot of good reasons why someone would want to provide a list of `args`, another good example is a program like `c8`, where you might provide options to both the program itself, and the file being instrumented:

```bash
c8 --reporter=foo myfile.js --some-argument-for-file
```

Going to go ahead and resolve this issue, let me know if there are any concerns :+1: